core/handler/downlink_test.go

```diff
@@ -94,4 +99,46 @@ func TestHandleDownlink(t *testing.T) {
 		Payload: []byte{96, 4, 3, 2, 1, 0, 1, 0, 1, 0, 0, 0, 0},
 	})
 	a.So(err, ShouldBeNil)
+
+	h.applications.Set(&application.Application{
+		AppID: appID,
+		Encoder: `function (payload){
+	  		return [96, 4, 3, 2, 1, 0, 1, 0, 1, 0, 0, 0, 0]
+			}`,
+	})
+	jsonFields := make(map[string]interface{})
+	jsonFields["key"] = "value"
+	jsonFields["key2"] = 11
+	// Both Payload and Fields provided : ERROR
+	err = h.HandleDownlink(&mqtt.DownlinkMessage{
+		FPort:   1,
+		AppID:   appID,
+		DevID:   devID,
+		Fields:  jsonFields,
+		Payload: []byte{0xAA, 0xBC},
+	}, &pb_broker.DownlinkMessage{
+		AppEui: &appEUI,
+		DevEui: &devEUI,
+	})
+	a.So(err, ShouldNotBeNil)
+
+	// Wait for the first goroutine to end
+	time.Sleep(time.Millisecond * 50)
```

Comment:
This can result in race conditions. Instead of waiting for the channel in a goroutine, try waiting after calling the method, possibly with a time-out (see https://blog.golang.org/go-concurrency-patterns-timing-out-and)